exporter/exporterhelper/tracehelper.go

```diff
@@ -86,57 +83,91 @@ func (p traceDataPusherOld) withObservability(exporterName string) traceDataPush
+	return newTracesRequest(req.ctx, partialErr.GetTraces(), req.pusher)
+}
+
+func (req *tracesRequest) export(ctx context.Context) (int, error) {
+	return req.pusher(ctx, req.td)
+}
+
+func (req *tracesRequest) count() int {
+	return req.td.SpanCount()
+}
+
 type traceExporter struct {
-	baseExporter
-	dataPusher traceDataPusher
+	*baseExporter
+	pusher traceDataPusher
 }
 
-func (te *traceExporter) ConsumeTraces(
-	ctx context.Context,
-	td pdata.Traces,
-) error {
-	exporterCtx := obsreport.ExporterContext(ctx, te.exporterFullName)
-	_, err := te.dataPusher(exporterCtx, td)
+func (texp *traceExporter) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	exporterCtx := obsreport.ExporterContext(ctx, texp.cfg.Name())
+	req := newTracesRequest(exporterCtx, td, texp.pusher)
+	_, err := texp.sender.send(req)
 	return err
 }
 
 // NewTraceExporter creates a TraceExporter that can record metrics and can wrap
 // every request with a Span.
 func NewTraceExporter(
-	config configmodels.Exporter,
+	cfg configmodels.Exporter,
 	dataPusher traceDataPusher,
 	options ...ExporterOption,
 ) (component.TraceExporter, error) {
 
-	if config == nil {
+	if cfg == nil {
 		return nil, errNilConfig
 	}
 
 	if dataPusher == nil {
 		return nil, errNilPushTraceData
 	}
 
-	dataPusher = dataPusher.withObservability(config.Name())
+	be := newBaseExporter(cfg, options...)
+
+	// Record metrics on the consumer.
+	be.qSender.nextSender = &tracesExporterWithObservability{
+		exporterName: cfg.Name(),
+		sender:       be.qSender.nextSender,
+	}
 
 	return &traceExporter{
-		baseExporter: newBaseExporter(config.Name(), options...),
-		dataPusher:   dataPusher,
+		baseExporter: be,
+		pusher:       dataPusher,
 	}, nil
 }
 
-// withObservability wraps the current pusher into a function that records
-// the observability signals during the pusher execution.
-func (p traceDataPusher) withObservability(exporterName string) traceDataPusher {
-	return func(ctx context.Context, td pdata.Traces) (int, error) {
-		ctx = obsreport.StartTraceDataExportOp(ctx, exporterName)
-		// Forward the data to the next consumer (this pusher is the next).
-		droppedSpans, err := p(ctx, td)
+type tracesExporterWithObservability struct {
+	exporterName string
+	sender       requestSender
+}
 
-		// TODO: this is not ideal: it should come from the next function itself.
-		// 	temporarily loading it from internal format. Once full switch is done
-		// 	to new metrics will remove this.
-		numSpans := td.SpanCount()
-		obsreport.EndTraceDataExportOp(ctx, numSpans, droppedSpans, err)
-		return droppedSpans, err
-	}
+func (tewo *tracesExporterWithObservability) send(req request) (int, error) {
+	req.setContext(obsreport.StartTraceDataExportOp(req.context(), tewo.exporterName))
+	// Forward the data to the next consumer (this pusher is the next).
+	droppedSpans, err := tewo.sender.send(req)
+
+	// TODO: this is not ideal: req should come from the next function itself.
```

Comment:
Copy/paste issues?

Reply:
Was a replace using goland for a variable which decided to replace `it` -> `req` :)